### PR TITLE
#171380759 Avail requests for approval

### PR DESCRIPTION
**Pull request by the22mastermind**
#### What does this PR do?
* Add `Approve trip request` functionality
* Add `Reject trip request` functionality
* Add `Reassign trip request` functionality
* Update user profile by setting line manager
#### Description of Task to be completed?
* Write endpoints unit tests
* Create controller functions
* Create helper functions, validations, middlewares, and custom messages.
* Create `approve`, `reject`, and `reassign` routes
#### How should this be manually tested?
1. Open your terminal and run `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend/`
2. run `cd stackup2-barefoot-backend` to go into the root directory of the project
3. run `git checkout -b ft-avail-requests-for-approval-171380759 origin/ft-avail-requests-for-approval-171380759` to pull this branch and switch to it
4. run `npm install` to install the dependencies
5. In .env file, add these variables:
SUPER_ADMIN_EMAIL = #superadmin@admin.com
SUPER_ADMIN_PASSWORD = #SuperUserPassword
* run `npm run migrations`
* run `npm run admin` to create a super admin account
* run `npm run dev`
6. Using POSTMAN, create 3 different accounts and verify them, then using super admin token, set one account 1 and 2 have a role of manager
7. Use the remaining account to create a trip request

**Approve a trip request**
1. As a requester, update your profile and set a line manager:
PATCH `http://localhost:<port>/api/profile`
Body request:
```
{
 lineManager: id
}
```
2. Login as the manager who just got a direct report above
3. Approve/reject your direct report's trip request: 
  PATCH `http://localhost:<port>/api/trips/<tripRequestId>/approve`

**Reject a trip request**
  PATCH `http://localhost:<port>/api/trips/<tripRequestId>/reject`

**Reassign a trip request to another manager**
  PATCH `http://localhost:<port>/api/trips/<tripRequestId>/reassign`
Body request:
```
{
  "userId": "id"
}
```
#### Any background context you want to provide?
* A manager can only approve or reject trip requests from their direct report and the ones assigned to him/her
* Only managers can reassign requests to other managers
* An already approved or rejected request cannot be reassigned
#### What are the relevant pivotal tracker stories?
[#171380759](https://www.pivotaltracker.com/story/show/171380759)
[#171380758](https://www.pivotaltracker.com/story/show/171380758)
#### Screenshots (if appropriate)
Screenshot from 2020-04-02 07-43-05
Screenshot from 2020-04-02 07-43-49
Screenshot from 2020-04-02 07-45-43

#### Questions:
N/A

### CODE REVIEW

File: src/controllers/request.controller.js
Comment (since9629) [inline src/controllers/request.controller.js]: could you use globals declared in src/utils/tripRequestsStatus.util.js

File: src/database/migrations/20200312122638-create-request.js
Comment (since9629) [inline src/database/migrations/20200312122638-create-request.js]: Are these people different `handledBy` & `handlerNames`? I think you don't need to add this field since you already referenced the handler in `handledBy` field.

File: src/routes/api/request.routes.js
Comment (since9629) [inline src/routes/api/request.routes.js:44]: why does it look like you added this code?

File: src/routes/api/request.routes.js
Comment (since9629) [inline src/routes/api/request.routes.js]: why does it look like you added this code?

File: src/routes/api/request.routes.js
Comment (the22mastermind) [reply src/routes/api/request.routes.js:44]: Code climate was saying the destructure after the imports has issues so I had to modify it like this to make it work.

File: src/routes/api/request.routes.js
Comment (the22mastermind) [reply src/routes/api/request.routes.js]: Same issue as above.

**Review by Muhire-Josue — CHANGES_REQUESTED** [review]
great job @the22mastermind however could you resolve those conflicts so that we can review your PR?

**Review by Ugizwenayo-Divine — COMMENTED** [review]
You did very good work 
You should rebase your work

**Review by since9629 — APPROVED** [review]

**Comment by kipropbrian** [comment]
Some comments just on reading the description.

- only your line manager should approve your requests, unless the requests are reassigned. 
- Instead of using `email` to identify the manager to reassign the request, use their ID's. Take into account this will work on the frontend. It will probably be a drop down with a list of managers from which we choose. The backend just needs the ID of that manager, not their email.

**Comment by the22mastermind** [comment]
> Some comments just on reading the description.
> 
> * only your line manager should approve your requests, unless the requests are reassigned.
> * Instead of using `email` to identify the manager to reassign the request, use their ID's. Take into account this will work on the frontend. It will probably be a drop down with a list of managers from which we choose. The backend just needs the ID of that manager, not their email.

@kipropbrian Thanks for the review.
1. Ok, I am going to work on it.
2. I'm going to change and use the ID. On the front-end I thought it could be like a modal with a search field where a manager can search by name or email in case it's a large company that has a lot of managers. Should I change the search field to a drop down as well?
Manager - Assign Request Modal

**Review by Ugizwenayo-Divine — APPROVED** [review]
Well done

**Review by kipropbrian — APPROVED** [review]

**Pull request merged**
